File: DesignPatternsExercises/ruby/memento_design_pattern/exercise/memento_design_pattern.rb
require_relative 'home_theatre'
require_relative 'led_tv'
require_relative 'speaker'
require_relative 'warehouse'

tv1 = LedTV.new(10, 100)
tv2 = LedTV.new(20, 200)

speaker1 = Speaker.new(1000)
speaker2 = Speaker.new(2000)

warehouse = Warehouse.new

home_theatre = HomeTheatre.new(tv1, speaker1)
warehouse.add_memento(home_theatre.memento)

home_theatre = HomeTheatre.new(tv1, speaker2)
warehouse.add_memento(home_theatre.memento)

home_theatre = HomeTheatre.new(tv2, speaker1)
warehouse.add_memento(home_theatre.memento)

home_theatre = HomeTheatre.new(tv2, speaker2)
warehouse.add_memento(home_theatre.memento)

puts "-- -- Current Home Theatre State is #{HomeTheatre.state_count - 1} -- --"
puts "Home Theatre State #{HomeTheatre.state_count - 1}:\n    #{home_theatre.to_string}"

puts '-- -- Rollback to Home Theatre State 0 -- --'
home_theatre.memento = warehouse.memento(0)
puts "Home Theatre State 0:\n    #{home_theatre.to_string}"

File: DesignPatternsExercises/java/MementoDesignPattern/Solution/MementoDesignPattern.java
public class MementoDesignPattern {
    public static void main(String[] args) {
        LedTV tv1 = new LedTV(10, 100);
        LedTV tv2 = new LedTV(20, 200);

        Speaker speaker1 = new Speaker(1000);
        Speaker speaker2 = new Speaker(2000);

        Warehouse warehouse = new Warehouse();

        HomeTheatre homeTheatre = new HomeTheatre(tv1, speaker1);
        warehouse.addMemento(homeTheatre.getMemento());

        homeTheatre = new HomeTheatre(tv1, speaker2);
        warehouse.addMemento(homeTheatre.getMemento());

        homeTheatre = new HomeTheatre(tv2, speaker1);
        warehouse.addMemento(homeTheatre.getMemento());

        homeTheatre = new HomeTheatre(tv2, speaker2);
        warehouse.addMemento(homeTheatre.getMemento());

        System.out.println("-- -- Current Home Theatre State is " + (HomeTheatre.getStateCount() - 1) + " -- --");
        System.out.println("Home Theatre State " + (HomeTheatre.getStateCount() - 1) + ":\n    " + homeTheatre);

        System.out.println("-- -- Rollback to Home Theatre State 0 -- --");
        homeTheatre.setMemento(warehouse.getMemento(0));
        System.out.println("Home Theatre State 0:\n    " + homeTheatre);
    }
}
